File: 10-GRAPHS/Go/clone-graph.go
package main

// https://leetcode.com/problems/clone-graph/

type Node struct {
	Val       int
	Neighbors []*Node
}

func cloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	visited := map[*Node]*Node{}

	var dfs func(node *Node) *Node
	dfs = func(node *Node) *Node {
		if v, ok := visited[node]; ok {
			return v
		}

		clone := &Node{Val: node.Val}
		visited[node] = clone
		for _, nei := range node.Neighbors {
			clone.Neighbors = append(clone.Neighbors, dfs(nei))
		}

		return clone
	}

	return dfs(node)
}

File: 10-GRAPHS/Python/clone-graph.py
from typing import Optional


# https://leetcode.com/problems/clone-graph/

class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node is None:
            return None

        visited = dict()

        def dfs(n: Node):
            if n in visited:
                return visited[n]

            clone = Node(val=n.val)
            visited[n] = clone

            for nei in n.neighbors:
                clone.neighbors.append(dfs(nei))

            return clone

        return dfs(node)

